terraform/amazon/ecs.tf

```diff
@@ -9,45 +10,49 @@ resource "aws_ecs_task_definition" "main" {
   task_role_arn      = data.aws_iam_role.task_container_role.arn
   execution_role_arn = data.aws_iam_role.task_execution_role.arn
 
-  container_definitions = <<DEFINITION
-[
-  {
-    "cpu": 0,
-    "image": "${data.aws_ecr_repository.main.repository_url}:latest",
-    "name": "${var.logical_name}",
-    "networkMode": "awsvpc",
-    "logConfiguration": {
-      "logDriver": "awslogs",
-      "options": {
-      "awslogs-group": "ecs/${var.logical_name}",
-      "awslogs-region": "${var.region}",
-      "awslogs-stream-prefix": "${var.logical_name}"
-      }
-    },
-    "healthCheck": {
-      "command": ["CMD-SHELL", "curl -f http://localhost:${var.port}${var.health_check_endpoint} || exit 1"],
-      "interval": 45,
-      "timeout" : 5,
-      "retries" : 3,
-      "startPeriod" : ${var.health_check_grace_period}
-    },
-    "portMappings": [
-      {
-        "containerPort": ${var.port},
-        "hostPort": ${var.port}
+  container_definitions = templatefile("${path.module}/task-definitions/main.tftpl", local.main_vars)
+}
+
+resource "aws_ecs_task_definition" "efs" {
+  count        = var.use_efs ? 1 : 0
+  family       = var.logical_name
+  network_mode = "awsvpc"
+  requires_compatibilities = [
+    "FARGATE",
+  ]
+  cpu                = var.cpu
+  memory             = var.memory
+  task_role_arn      = data.aws_iam_role.task_container_role.arn
+  execution_role_arn = data.aws_iam_role.task_execution_role.arn
+
+  container_definitions = templatefile("${path.module}/task-definitions/main-efs.tftpl", local.main_efs_vars)
+
+  volume {
+    name = var.efs_name
+    efs_volume_configuration {
+      file_system_id          = var.efs_file_system_id
+      root_directory          = var.efs_path
+      transit_encryption      = "ENABLED"
+      transit_encryption_port = 2049
+      authorization_config {
+        access_point_id   = var.efs_access_point_id
+        iam               = "ENABLED"
       }
-    ]
+    }
   }
-]
-DEFINITION
+}
 
+locals {
+  aws_ecs_task_definition_arn = {
+    value = element(concat(aws_ecs_task_definition.main.*.arn, aws_ecs_task_definition.efs.*.arn, list("")), 0)
```

Comment:
There's likely a reason why doing this is cumbersome in terraform. It may very well be by design and not necessarily a limitation of terraform.

Since terraforms are generally reflective of the state of the AWS or GCP environment, introducing conditions and forks for resource provisioning might be considered an anti-pattern. 

There is also an option to change the deployment path to be executed directly from Github actions too, which would put the onus on the creating service to deploy a task definition to AWS.

https://github.com/aws-actions/amazon-ecs-deploy-task-definition

Reply:
The scope of this change becomes larger in this respect, but would dive into the paradigm of using terraform strictly for resource provisioning and pass the CI/CD (Deployments) to GitHub.